starlite/router.py

```diff
@@ -223,32 +221,49 @@ def route_handler_method_view(
+    RouteHandlerMap = ItemsView[
+        str, Union[WebsocketRouteHandler, ASGIRouteHandler, Dict["HttpMethod", HTTPRouteHandler]]
+    ]
+
+    @classmethod
     def _map_route_handlers(
+        cls,
         value: Union[Controller, BaseRouteHandler, "Router"],
-    ) -> ItemsView[str, Union[WebsocketRouteHandler, ASGIRoute, Dict["HttpMethod", HTTPRouteHandler]]]:
-        """Maps route handlers to http methods."""
-        handlers_map: Dict[str, Any] = {}
-        if isinstance(value, BaseRouteHandler):
-            for path in value.paths:
-                if isinstance(value, HTTPRouteHandler):
-                    handlers_map[path] = {http_method: value for http_method in value.http_methods}
-                elif isinstance(value, (WebsocketRouteHandler, ASGIRouteHandler)):
-                    handlers_map[path] = value
-        elif isinstance(value, Router):
+    ) -> RouteHandlerMap:
+        """Maps route handlers to HTTP methods."""
+        if isinstance(value, Router):
+            handlers_map: Dict[str, Any] = {}
             handlers_map = value.route_handler_method_map
-        else:
-            # we are dealing with a controller
-            for route_handler in value.get_route_handlers():
-                for handler_path in route_handler.paths:
-                    path = join_paths([value.path, handler_path]) if handler_path else value.path
-                    if isinstance(route_handler, HTTPRouteHandler):
-                        if not isinstance(handlers_map.get(path), dict):
-                            handlers_map[path] = {}
-                        for http_method in route_handler.http_methods:
-                            handlers_map[path][http_method] = route_handler
-                    else:
-                        handlers_map[path] = cast("Union[WebsocketRouteHandler, ASGIRouteHandler]", route_handler)
+            return handlers_map.items()
+        if isinstance(value, BaseRouteHandler):
+            return cls._map_route_handlers_for_base_route_handler(value)
+        # we are dealing with a controller
+        return cls._map_route_handlers_for_controller(value)
+
+    @staticmethod
+    def _map_route_handlers_for_base_route_handler(value: BaseRouteHandler) -> RouteHandlerMap:
+        """Maps route handlers to HTTP methods for an input
+        BaseRouteHandler."""
+        handlers_map: Dict[str, Any] = {}
+        for path in value.paths:
+            if isinstance(value, HTTPRouteHandler):
+                handlers_map[path] = {http_method: value for http_method in value.http_methods}
+            elif isinstance(value, (WebsocketRouteHandler, ASGIRouteHandler)):
+                handlers_map[path] = value
+        return handlers_map.items()
+
+    @staticmethod
+    def _map_route_handlers_for_controller(value: Controller) -> RouteHandlerMap:
+        """Maps route handlers to HTTP methods for an input Controller."""
+        handlers_map: Dict[str, Any] = collections.defaultdict(dict)
```

Comment:
why a defaultdict?

Reply:
Because it avoids the check to see whether the key already exists in the dict - it removes the need for the first two lines here:

```
if not isinstance(handlers_map.get(path), dict):
    handlers_map[path] = {}
for http_method in route_handler.http_methods:
    handlers_map[path][http_method] = route_handler
```